Natural Language and Image-Based Search Support for Recordings

WANGXIAOMIN-HIK opened the pull request:
To enhance ONVIF's search capabilities, the following operations have been added to support natural language and image-based search for video recordings:

FindImagebyNL
Purpose: Starts a search session using natural language descriptions to locate relevant video recordings. Example Query: "Person wearing a red hat."
Parameters:
StartPoint: Start time for the search.
EndPoint: End time for the search.
RecordingToken: (Optional) Token for the recording to search. Text: Natural language description for the search. Likelihood: (Optional) Similarity threshold for the search (0~1). MaxMatches: (Optional) Maximum number of matches to return. KeepAliveTime: Time the search session will be kept alive. Response:
SearchToken: A unique reference to the search session. GetNLSearchResults
Purpose: Retrieves results from a natural language search session initiated by FindImagebyNL. Parameters:
SearchToken: Token identifying the search session. MinResults: (Optional) Minimum number of results to return. MaxResults: (Optional) Maximum number of results to return. WaitTime: (Optional) Maximum time to wait for results. Response:
ResultList: List of matching results, including metadata such as TargetImageURI, Time, Likelihood, and RecordingToken. FindImagebyImage
Purpose: Starts a search session using a target image to locate relevant video recordings. Parameters:
StartPoint: Start time for the search.
EndPoint: End time for the search.
RecordingToken: (Optional) Token for the recording to search. TargetImageURI: URI of the target image to be searched. MaxMatches: (Optional) Maximum number of matches to return. KeepAliveTime: Time the search session will be kept alive. Response:
SearchToken: A unique reference to the search session. GetImageSearchResults
Purpose: Retrieves results from an image-based search session initiated by FindImagebyImage. Parameters:
SearchToken: Token identifying the search session. MinResults: (Optional) Minimum number of results to return. MaxResults: (Optional) Maximum number of results to return. WaitTime: (Optional) Maximum time to wait for results. Response:
ResultList: List of matching results, including metadata such as TargetImageURI, Time, Likelihood, and RecordingToken. Schema Updates
onvif.xsd:

Added complex types for FindImageResult and FindImageResultList to support result structures for both natural language and image-based searches. Included fields like TargetImageURI, Time, Likelihood, and RecordingToken. search.wsdl:

Defined operations FindImagebyNL, GetNLSearchResults, FindImagebyImage, and GetImageSearchResults. Added request and response elements for each operation. Documentation Updates
RecordingSearch.xml:
Added detailed descriptions for FindImagebyNL and GetNLSearchResults operations, explaining their purpose, parameters, and responses.

WANGXIAOMIN-HIK committed:
Natural Language and Image-Based Search Support for Recordings
To enhance ONVIF's search capabilities, the following operations have been added to support natural language and image-based search for video recordings:

FindImagebyNL
Purpose: Starts a search session using natural language descriptions to locate relevant video recordings.
Example Query: "Person wearing a red hat."
Parameters:
StartPoint: Start time for the search.
EndPoint: End time for the search.
RecordingToken: (Optional) Token for the recording to search.
Text: Natural language description for the search.
Likelihood: (Optional) Similarity threshold for the search (0~1).
MaxMatches: (Optional) Maximum number of matches to return.
KeepAliveTime: Time the search session will be kept alive.
Response:
SearchToken: A unique reference to the search session.
GetNLSearchResults
Purpose: Retrieves results from a natural language search session initiated by FindImagebyNL.
Parameters:
SearchToken: Token identifying the search session.
MinResults: (Optional) Minimum number of results to return.
MaxResults: (Optional) Maximum number of results to return.
WaitTime: (Optional) Maximum time to wait for results.
Response:
ResultList: List of matching results, including metadata such as TargetImageURI, Time, Likelihood, and RecordingToken.
FindImagebyImage
Purpose: Starts a search session using a target image to locate relevant video recordings.
Parameters:
StartPoint: Start time for the search.
EndPoint: End time for the search.
RecordingToken: (Optional) Token for the recording to search.
TargetImageURI: URI of the target image to be searched.
MaxMatches: (Optional) Maximum number of matches to return.
KeepAliveTime: Time the search session will be kept alive.
Response:
SearchToken: A unique reference to the search session.
GetImageSearchResults
Purpose: Retrieves results from an image-based search session initiated by FindImagebyImage.
Parameters:
SearchToken: Token identifying the search session.
MinResults: (Optional) Minimum number of results to return.
MaxResults: (Optional) Maximum number of results to return.
WaitTime: (Optional) Maximum time to wait for results.
Response:
ResultList: List of matching results, including metadata such as TargetImageURI, Time, Likelihood, and RecordingToken.
Schema Updates
onvif.xsd:

Added complex types for FindImageResult and FindImageResultList to support result structures for both natural language and image-based searches.
Included fields like TargetImageURI, Time, Likelihood, and RecordingToken.
search.wsdl:

Defined operations FindImagebyNL, GetNLSearchResults, FindImagebyImage, and GetImageSearchResults.
Added request and response elements for each operation.
Documentation Updates
RecordingSearch.xml:
Added detailed descriptions for FindImagebyNL and GetNLSearchResults operations, explaining their purpose, parameters, and responses.

WANGXIAOMIN-HIK committed:
Expand the recording search feature to support multi-recording token search
Updated document and WSDL definitions to allow multiple recording tokens to be passed in in a search operation to query multiple recordings at the same time.

venki5685 commented on wsdl/ver10/search.wsdl:
```diff
@@ -444,7 +446,143 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
+						<xs:element name="EndPoint" type="xs:dateTime">
+							<xs:annotation><xs:documentation>End time for the search.</xs:documentation></xs:annotation>
+						</xs:element>
+						<xs:element name="RecordingToken" type="tt:RecordingReference" minOccurs="0" maxOccurs="unbounded">
```
add maxOccurs="unbounded" which will allow more than one recording container to search.

venki5685 commented on wsdl/ver10/search.wsdl:
```diff
@@ -444,7 +446,143 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
+			</xs:element>
+
+			<!-- Define FindImagebyImage -->
+			<xs:element name="FindImagebyImageRequest">
```
FindImagebyImage name can be revisited.

WANGXIAOMIN-HIK replied:
Yes, thank you for your opinion. We change the FindImageByImage to SearchImageByImage, and FindImageByNL to SearchImageByNL.

WANGXIAOMIN-HIK committed:
#603 update
1.change the FindImageByImage to SearchImageByImage, and FindImageByNL to SearchImageByNL.
2.memo the TargetImageURI is the result returned from SearchImageByNL.

kieran242 commented:
@WANGXIAOMIN-HIK  

Is this functionality aimed at a Camera device or Network Video Recorder or Both?
How is the device supporting NL?

WANGXIAOMIN-HIK commented on wsdl/ver10/search.wsdl:
```diff
@@ -444,7 +446,143 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
+							<xs:annotation><xs:documentation>This element contains a list of recording tokens to search.</xs:documentation></xs:annotation>
+						</xs:element>
+						<xs:element name="TargetImageURI" type="xs:anyURI">
+						<xs:annotation><xs:documentation>The target image to be searched in LocalStorage URI format.</xs:documentation></xs:annotation>
```
thank you for your opinion. @venki5685 @kieran242 
We think about it carefully, SearchImagebyImage and SearchImageByNL are independent search sessions.
TargetImageURI can be a local URI or a remote URI.

WANGXIAOMIN-HIK commented:
Yes, thank you for your opinion. @kieran242 

Yes, Both. this functionality aimed at a Camera device and Network Video Recorder.
To implement this function on the camera device, the camera device needs to have storage space for images, such as installing a SD card.

The device implements the following algorithm，algorithm leverages massive annotated image-text pairs during training, where visual features (e.g., "dog", "snow", "yellow” ,“fur") and textual elements (e.g., tokenized "snow/field/dog") are extracted through cross-modal neural networks, forming the foundation of its text-to-image retrieval model. In deployment, the system processes video streams to detect targets, then employs on-device models to generate and store binary-encoded feature vectors for rapid matching.

kieran242 commented:
@WANGXIAOMIN-HIK very kind thanks for your response. It was very informative.

kieran242 reviewed (commented):
@WANGXIAOMIN-HIK  a few minor updates as suggestions wsdl update is good but spelling mistake in doc.

WANGXIAOMIN-HIK committed:
Correct the spelling mistake : SerachImagebyImage -> SearchImagebyImage

WANGXIAOMIN-HIK commented on doc/RecordingSearch.xml:
```diff
@@ -1320,6 +1320,193 @@ http://www.onvif.org/ver10/tptz/ZoomSpaces/PositionGenericSpace
+    </section>    
+    <section>
+      <title>GetImageSearchResults</title>
+      <para>GetImageSearchResults acquires the results from an image-based search session previously initiated by a SerachImagebyImage operation. The response shall not include results already returned in previous requests for the same session.</para>
```
Thank you very much for your advice. I have fixed the spelling error. I appreciate your help.

WANGXIAOMIN-HIK committed:
Verander SearchImagebyImage na SearchImageByImage

venki5685 commented on wsdl/ver10/schema/onvif.xsd:
```diff
@@ -7621,6 +7627,37 @@ and sample rate. </xs:documentation>
+	<!-- FindImageResult -->
+	<xs:complexType name="FindImageResult">
+		<xs:sequence>
+			<xs:element name="TargetImageURI" type="xs:anyURI">
```
is this target image uri maps to scene image or target object image?

WANGXIAOMIN-HIK replied:
object image.
The URI of the detected target object image. This can be either:
1、a local image stored in the NPL Target Image repository (LocalStorage format).
2、an external image provided by the client for image search or feature matching.

dstafx replied:
Consider renaming FindImageResult to FindObjectImageResult or FindObjectSnapshotResult to make it more clear that images of object are returned and not images of the whole scene.

venki5685 commented on doc/RecordingSearch.xml:
```diff
@@ -1320,6 +1320,193 @@ http://www.onvif.org/ver10/tptz/ZoomSpaces/PositionGenericSpace
+            <para role="text">The point of time where the search will stop.</para>
+            <para role="param">RecordingToken - optional [tt:RecordingReference]</para>
+            <para role="text">Token for the recording to search.</para>
+            <para role="param">TargetImageURI [xs:anyURI]</para>
```
add a search parameter to accept external image from client in addition to NPL Target image URI. Either client can use NPLTargetImageURI or External Image from client for image search feature.

WANGXIAOMIN-HIK committed:
Update the search functionality, add the TargetImageData parameter, and provide detailed explanation of the use of TargetImageURI

WANGXIAOMIN-HIK committed:
Update the TargetImageURI description, remove the NPL prefix to unify the terminology, as the original meaning refers to the images stored internally on the device.

WANGXIAOMIN-HIK committed:
change Likelihood score to Likelihood threshold , change FindImageResult to FindObjectImageResult.

WANGXIAOMIN-HIK commented on wsdl/ver10/schema/onvif.xsd:
```diff
@@ -7621,6 +7627,37 @@ and sample rate. </xs:documentation>
+			<xs:element name="Time" type="xs:dateTime">
+				<xs:annotation><xs:documentation>The time when the target was found. Users can use this as a reference point to search records that occurred before and after this specific time point.</xs:documentation></xs:annotation>
+			</xs:element>
+			<xs:element name="Likelihood" type="xs:float">
```
Thank you for your clear feedback. we change Likelihood score to Likelihood threshold. Represents similarity, of type float, ranging from 0 to 1; the higher the value, the higher the similarity.

WANGXIAOMIN-HIK replied:
vector distance

WANGXIAOMIN-HIK replied:
Would it be more appropriate to define it as CosineSimilarity? It reflects the way Cosine Distance is calculated.

WANGXIAOMIN-HIK commented on doc/RecordingSearch.xml:
```diff
@@ -1320,6 +1320,195 @@ http://www.onvif.org/ver10/tptz/ZoomSpaces/PositionGenericSpace
+      </variablelist>
+    </section>    
+    <section>
+      <title>GetImageSearchResults</title>
```
Thank you for your clear feedback. GetNLSearchResults and GetImageSearchResults are independent functions, and their request parameters differ significantly. We recommend maintaining the correspondence between the interfaces to make it easier for users to understand during integration.

WANGXIAOMIN-HIK committed:
Change the "Likelihood" field to "CosineSimilarity".
the description :It represents the cosine similarity between two vectors, which is used to measure the similarity of the directions of the vectors. The closer the value is to 1, the higher the similarity; the closer the value is to 0, the lower the similarity.

HansBusch commented on doc/RecordingSearch.xml:
```diff
@@ -1320,6 +1320,195 @@ http://www.onvif.org/ver10/tptz/ZoomSpaces/PositionGenericSpace
+            <para role="text">Token for the recording to search.</para>
+            <para role="param">Text [xs:string]</para>
+            <para role="text">Natural language description for the search.</para>
+            <para role="param">CosineSimilarity - optional [xs:float]</para>
```
CosineSimularity is just one out of a set of common simularity measures. 
I can imagine that ONVIF just defines an abstract simularity bewteen zero and one or a complex item supporting multiple similarity measures.

For the sake of simplicity I prefer the first approach as the second one would require a set of capabilities which ones a device supports.

WANGXIAOMIN-HIK replied:
yes, 
Cosine similarity is just one of the most commonly used similarity measures in vector space. In actual image retrieval/similarity assessment, multiple distance measures, matching based on local features, perceptual similarity, as well as learned metrics or hash/quantization indexing are also used.
Could we consider changing the field back to 'simularity', but in the comments, we can note that we are currently using the cosine vector method? @dstafx

dstafx replied:
Ok for me since this is a big and important topic that we likely need to come back to on a more general level. But to explain a bit more:

My general concern (and the industry challenge) is that if we are too generic it will not be useful across vendors as the implementation scores would not be compareable. The message to the client if it is a generic number is likely that every enpoint where this interface is offered may have different implementations so the similarity is not comparable between different endpoints. So when searching some devices or vendors may consistently report higher similarities thereby potentially hiding relevant results. By stating that the similarity is only for sorting search results from a single endpoint we can avoid this.

WANGXIAOMIN-HIK replied:
Thank you very much for your detailed explanation. After careful consideration, I still think it should be defined as similarity, as we cannot restrict the vendors' implementation algorithms.  

However, regarding the issue you mentioned about cross-device and cross-vendor search, we can add a note stating that this similarity is only guaranteed to be effective for ranking within the same search result returned by the same endpoint, and it does not guarantee that similarity can be compared across different vendors, devices, or endpoints.  

As for the cross-device search issue, we can discuss it in our next meeting. This would require imposing constraints on the hardware vendors' implementation mechanisms, such as ensuring that devices support the same algorithm scheduling to guarantee consistency in device detection mechanisms.

kieran242 commented:
@WANGXIAOMIN-HIK  @dstafx  @HansBusch is this issue resolved regarding the "similarity measures" ? I see it is required for IPR| review.

WANGXIAOMIN-HIK committed:
Change the "CosineSimilarity" field to "Similarity"

kieran242 reviewed (approved):
@WANGXIAOMIN-HIK approved with discussion on VE WG Call.

sujithhanwha commented:
It appears that both APIs, **GetImageSearchResults** and **GetNLSearchResults**, currently return identical results.

We could either consolidate them into a **single unified method** for search results, 
**or,**
if we prefer to keep them separate, define a distinct structure for NL search (e.g., **FindObjectImageResultList** → **FindNLResultList**) and adjust the response format to differentiate them. For instance, in **GetNLSearchResults**, the Image field could be optional when the search is based purely on metadata/text inference.

WANGXIAOMIN-HIK commented:
GetNLSearchResults and GetImageSearchResults are distinguished by their usage scenarios. 
GetNLSearchResults is used for searching images based on natural language, retrieving images that closely match the natural language description. 
GetImageSearchResults, on the other hand, searches for images using an image as input. It is based on image modeling and can accurately find the corresponding image information. 
Users first use GetNLSearchResults for preliminary retrieval, then select the images they want, and use GetImageSearchResults for precise searching. 
These two interfaces essentially both search for images, so they adopt the same structure.
@sujithhanwha

sujithhanwha commented:
> GetNLSearchResults and GetImageSearchResults are distinguished by their usage scenarios. GetNLSearchResults is used for searching images based on natural language, retrieving images that closely match the natural language description. GetImageSearchResults, on the other hand, searches for images using an image as input. It is based on image modeling and can accurately find the corresponding image information. Users first use GetNLSearchResults for preliminary retrieval, then select the images they want, and use GetImageSearchResults for precise searching. These two interfaces essentially both search for images, so they adopt the same structure. @sujithhanwha

@WANGXIAOMIN-HIK ,
My question is not about the usage scenarios—I fully agree with those distinctions. It’s purely about the interface design. If both GetImageSearchResults and GetNLSearchResults return the same format, why do we need two separate methods? Could we not unify them into a single method that accepts search token from both image and NL search.

If you believe it makes sense to keep them separate, I’d recommend using different result formats. This would allow each interface to evolve independently. (For example, if we later add parameters specific to image-based search, they wouldn’t automatically apply to natural language search results. )

WANGXIAOMIN-HIK committed:
We have made the following changes considering the future scalability of the interface.
We have added FindNLSearchResultList and FindNLSearchResult to distinguish the return values of the GetNLSearchResults and GetImageSearchResults interfaces.
GetNLSearchResults -> FindNLSearchResultList, FindNLSearchResult
GetImageSearchResults -> FindObjectImageResultList, FindObjectImageResult

WANGXIAOMIN-HIK commented:
Thank you for your suggestion. @sujithhanwha 
We have made the following changes considering the future scalability of the interface.
We have added FindNLSearchResultList and FindNLSearchResult to distinguish the return values of the GetNLSearchResults and GetImageSearchResults interfaces.
GetNLSearchResults -> FindNLSearchResultList, FindNLSearchResult
GetImageSearchResults -> FindObjectImageResultList, FindObjectImageResult

WeimingMaoDH commented on doc/RecordingSearch.xml:
```diff
@@ -1320,6 +1320,195 @@ http://www.onvif.org/ver10/tptz/ZoomSpaces/PositionGenericSpace
+            <para role="text">The point of time where the search will stop.</para>
+            <para role="param">RecordingToken - optional [tt:RecordingReference]</para>
+            <para role="text">Token for the recording to search.</para>
+            <para role="param">Text [xs:string]</para>
```
Whether the Text is a mandatory parameter? When Text is not provided in request, should device return ter:InvalidArgVal, or search for all ?  
And, Endpiont could be optional in my opinion.

WeimingMaoDH commented on doc/RecordingSearch.xml:
```diff
@@ -1320,6 +1320,195 @@ http://www.onvif.org/ver10/tptz/ZoomSpaces/PositionGenericSpace
+            <para role="param">RecordingToken - optional [tt:RecordingReference]</para>
+            <para role="text">Token for the recording to search.</para>
+            <para role="param">TargetImageURI - optional [xs:anyURI]</para>
+            <para role="text">The URI of the detected target object image. This can be either: - a local image stored in the Target Image repository (LocalStorage format), or - an external image provided by the client for image search or feature matching.</para>
```
Same quesion here. If neither of the two optional parameters TargetImageURI and TargetImageData are provided,  what is the expected behavior of the device?

WANGXIAOMIN-HIK committed:
Update the search function documentation to clarify that the EndPoint parameter is optional, and add instructions for searching the latest data; adjust TargetImageURI and TargetImageData to be mutually exclusive options.

WANGXIAOMIN-HIK committed:
Modify the field description of the EndPoint

WeimingMaoDH reviewed (approved):
Prototyping has completed with Dahua device. Interfaces work as expected.

WANGXIAOMIN-HIK committed:
Update the search function documentation, modify the description of the EndPoint parameter to clarify that it is optional, and adjust TargetImageURI and TargetImageData to be mutually exclusive options.

kieran242 reviewed (approved):
@WANGXIAOMIN-HIK re-reviewed changes and I am happy to apporove.

WANGXIAOMIN-HIK committed:
Merge branch 'development' of https://github.com/onvif/specs into video/NLSearch&ImageSearch
